File: test/cc/redpen/intellij/SettingsPaneTest.java
package cc.redpen.intellij;

import cc.redpen.config.*;
import com.google.common.collect.ImmutableMap;
import org.junit.Before;
import org.junit.Test;

import javax.swing.*;
import javax.swing.table.DefaultTableModel;
import javax.swing.table.TableModel;
import java.awt.*;
import java.io.File;
import java.io.FileOutputStream;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static cc.redpen.config.SymbolType.*;
import static java.util.Arrays.asList;
import static java.util.Collections.emptyMap;
import static java.util.Collections.singletonList;
import static javax.swing.JFileChooser.APPROVE_OPTION;
import static javax.swing.JFileChooser.CANCEL_OPTION;
import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

public class SettingsPaneTest extends BaseTest {
  RedPenProvider provider = new RedPenProvider(new LinkedHashMap<>(ImmutableMap.of("en", cloneableConfig("en"), "ja", cloneableConfig("ja"))));
  SettingsPane settingsPane = spy(new SettingsPane(provider));

  @Before
  public void setUp() throws Exception {
    settingsPane.validators = mock(JTable.class, RETURNS_DEEP_STUBS);
    settingsPane.symbols = mock(JTable.class, RETURNS_DEEP_STUBS);
  }

  @Test
  public void activeConfigIsClonedOnCreation() throws Exception {
    settingsPane.initLanguages();
    assertSame(provider.getActiveConfig().clone(), settingsPane.getConfig());
  }

  @Test
  public void autodetectCheckboxIsInitializedToFalse() throws Exception {
    provider.setAutodetect(false);
    settingsPane.initLanguages();
    assertFalse(settingsPane.autodetectLanguage.isSelected());
  }

  @Test
  public void autodetectCheckboxIsInitializedToTrue() throws Exception {
    provider.setAutodetect(true);
    settingsPane.initLanguages();
    assertTrue(settingsPane.autodetectLanguage.isSelected());
  }

  @Test
  public void languagesAndVariantsArePrepopulated() throws Exception {
    provider.setActiveConfig(provider.getConfig("ja"));

    settingsPane.initLanguages();

    assertEquals(2, settingsPane.language.getItemCount());
    assertEquals("en", settingsPane.language.getItemAt(0));
    assertEquals("ja", settingsPane.language.getItemAt(1));
    assertEquals("ja", settingsPane.language.getSelectedItem());
  }

  @Test
  public void getPaneInitsEverything() throws Exception {
    settingsPane = mock(SettingsPane.class);

    doCallRealMethod().when(settingsPane).getPane();
    doCallRealMethod().when(settingsPane).initTabs();

    settingsPane.getPane();

    verify(settingsPane).initLanguages();
    verify(settingsPane).initValidators();
    verify(settingsPane).initSymbols();
  }

  @Test
  public void changingOfLanguageAppliesOldChangesAndInitsNewValidatorsAndSymbols() throws Exception {
    doNothing().when(settingsPane).initTabs();
    doNothing().when(settingsPane).applyLocalChanges();

    settingsPane.getPane();
    assertSame(provider.getActiveConfig().clone(), settingsPane.getConfig());
    verify(settingsPane).initTabs();

    settingsPane.language.firePopupMenuWillBecomeVisible();
    verify(settingsPane).applyLocalChanges();

    settingsPane.language.setSelectedItem("ja");
    assertSame(provider.getConfig("ja").clone(), settingsPane.getConfig());
    verify(settingsPane, times(2)).initTabs();
  }

  @Test
  public void validatorsAreListedInSettings() throws Exception {
    List<ValidatorConfiguration> allValidators = asList(
      validatorConfig("first", ImmutableMap.of("attr1", "val1", "attr2", "val2")),
      validatorConfig("second one", emptyMap()));

    when(provider.getInitialConfig("en").getValidatorConfigs()).thenReturn(allValidators);
    doReturn(configWithValidators(singletonList(validatorConfig("second one", emptyMap())))).when(settingsPane).getConfig();

    DefaultTableModel model = mock(DefaultTableModel.class);
    doReturn(model).when(settingsPane).createValidatorsModel();

    settingsPane.initValidators();

    verify(model).addRow(new Object[] {false, "first", "attr2=val2, attr1=val1"});
    verify(model).addRow(new Object[] {true, "second one", ""});
  }

  @Test
  public void getActiveValidators_returnsOnlySelectedValidators() throws Exception {
    settingsPane.initLanguages();
    when(provider.getInitialConfig("en").getValidatorConfigs()).thenReturn(asList(
      new ValidatorConfiguration("first"),
      new ValidatorConfiguration("second one")));

    when(settingsPane.validators.getModel().getRowCount()).thenReturn(2);
    when(settingsPane.validators.getModel().getValueAt(0, 0)).thenReturn(false);
    when(settingsPane.validators.getModel().getValueAt(1, 0)).thenReturn(true);
    when(settingsPane.validators.getModel().getValueAt(1, 2)).thenReturn("");

    List<ValidatorConfiguration> activeValidators = settingsPane.getActiveValidators();
    assertEquals(1, activeValidators.size());
    assertEquals("second one", activeValidators.get(0).getConfigurationName());
  }

  @Test
  public void getActiveValidators_modifiesAttributes() throws Exception {
    settingsPane.initLanguages();
    when(provider.getInitialConfig("en").getValidatorConfigs()).thenReturn(
      singletonList(validatorConfig("Hello", ImmutableMap.of("width", "100", "height", "300", "depth", "1"))));

    when(settingsPane.validators.getModel().getRowCount()).thenReturn(1);
    when(settingsPane.validators.getModel().getValueAt(0, 0)).thenReturn(true);
    when(settingsPane.validators.getModel().getValueAt(0, 2)).thenReturn(" width=200 ,   height=300 ");

    List<ValidatorConfiguration> activeValidators = settingsPane.getActiveValidators();
    assertEquals(1, activeValidators.size());
    assertEquals(ImmutableMap.of("width", "200", "height", "300"), activeValidators.get(0).getAttributes());
    assertNotSame(provider.getInitialConfig("en").getValidatorConfigs().get(0), activeValidators.get(0));
  }

  @Test
  public void getActiveValidators_reportsInvalidAttributes() throws Exception {
    settingsPane.initLanguages();
    ValidatorConfiguration validator = validatorConfig("Hello", ImmutableMap.of("width", "100"));
    when(provider.getInitialConfig("en").getValidatorConfigs()).thenReturn(singletonList(validator));

    doNothing().when(settingsPane).showPropertyError(any(ValidatorConfiguration.class), anyString());

    when(settingsPane.validators.getModel().getRowCount()).thenReturn(1);
    when(settingsPane.validators.getModel().getValueAt(0, 0)).thenReturn(true);

    when(settingsPane.validators.getModel().getValueAt(0, 2)).thenReturn("width");
    settingsPane.getActiveValidators();
    verify(settingsPane).showPropertyError(validator, "width");

    when(settingsPane.validators.getModel().getValueAt(0, 2)).thenReturn("=");
    settingsPane.getActiveValidators();
    verify(settingsPane).showPropertyError(validator, "=");
  }

  @Test
  public void getActiveValidators_appliesActiveCellEditorChanges() throws Exception {
    when(settingsPane.validators.isEditing()).thenReturn(true);
    settingsPane.getActiveValidators();
    verify(settingsPane.validators.getCellEditor()).stopCellEditing();
  }

  @Test
  public void getSymbols_appliesActiveCellEditorChanges() throws Exception {
    when(settingsPane.symbols.isEditing()).thenReturn(true);
    settingsPane.getSymbols();
    verify(settingsPane.symbols.getCellEditor()).stopCellEditing();
  }

  @Test
  public void symbolsAreListedInSettings() throws Exception {
    settingsPane.setConfig(configWithSymbols(asList(new Symbol(AMPERSAND, '&', "$%", true, false), new Symbol(ASTERISK, '*', "", false, true))));

    DefaultTableModel model = mock(DefaultTableModel.class);
    doReturn(model).when(settingsPane).createSymbolsModel();

    settingsPane.initSymbols();

    verify(model).addRow(new Object[] {AMPERSAND.toString(), '&', "$%", true, false});
    verify(model).addRow(new Object[] {ASTERISK.toString(), '*', "", false, true});
  }

  @Test
  public void getSymbols() throws Exception {
    TableModel model = settingsPane.symbols.getModel();
    when(model.getRowCount()).thenReturn(2);

    when(model.getValueAt(0, 0)).thenReturn("AMPERSAND");
    when(model.getValueAt(0, 1)).thenReturn('&');
    when(model.getValueAt(0, 2)).thenReturn("$%");
    when(model.getValueAt(0, 3)).thenReturn(true);
    when(model.getValueAt(0, 4)).thenReturn(false);

    when(model.getValueAt(1, 0)).thenReturn("ASTERISK");
    when(model.getValueAt(1, 1)).thenReturn("*");
    when(model.getValueAt(1, 2)).thenReturn("");
    when(model.getValueAt(1, 3)).thenReturn(false);
    when(model.getValueAt(1, 4)).thenReturn(true);

    List<Symbol> symbols = settingsPane.getSymbols();
    assertEquals(asList(new Symbol(AMPERSAND, '&', "$%", true, false), new Symbol(ASTERISK, '*', "", false, true)), symbols);
  }

  @Test
  public void fileChooserUsesXmlFileFilter() throws Exception {
    assertEquals("RedPen Configuration", settingsPane.fileChooser.getFileFilter().getDescription());
    File file = mock(File.class);

    when(file.getName()).thenReturn("blah.xml");
    assertTrue(settingsPane.fileChooser.getFileFilter().accept(file));

    when(file.getName()).thenReturn("blah.txt");
    assertFalse(settingsPane.fileChooser.getFileFilter().accept(file));
  }

  @Test
  public void canCancelExportingConfiguration() throws Exception {
    prepareImportExport();
    settingsPane.initButtons();
    when(settingsPane.fileChooser.showSaveDialog(any(Component.class))).thenReturn(CANCEL_OPTION);

    settingsPane.exportButton.doClick();

    verify(settingsPane, never()).apply();
    verify(settingsPane.fileChooser).showSaveDialog(settingsPane.root);
    verifyNoMoreInteractions(settingsPane.fileChooser);
  }

  @Test
  public void canExportConfiguration() throws Exception {
    prepareImportExport();
    File file = File.createTempFile("redpen-conf", ".xml");
    file.deleteOnExit();

    when(settingsPane.fileChooser.showSaveDialog(any(Component.class))).thenReturn(APPROVE_OPTION);
    when(settingsPane.fileChooser.getSelectedFile()).thenReturn(file);

    settingsPane.exportConfig();

    verify(settingsPane).apply();
    verify(settingsPane.fileChooser).showSaveDialog(settingsPane.root);
    verify(settingsPane.fileChooser).getSelectedFile();
    verify(settingsPane.configurationExporter).export(eq(settingsPane.getConfig()), any(FileOutputStream.class));
  }

  @Test
  public void canCancelImportingConfiguration() throws Exception {
    prepareImportExport();
    settingsPane.initButtons();
    when(settingsPane.fileChooser.showOpenDialog(any(Component.class))).thenReturn(CANCEL_OPTION);

    settingsPane.importButton.doClick();

    verify(settingsPane.fileChooser).showOpenDialog(settingsPane.root);
    verifyNoMoreInteractions(settingsPane.fileChooser);
  }

  @Test
  public void canImportConfiguration() throws Exception {
    prepareImportExport();
    File file = mock(File.class);
    Configuration config = config("ja.hankaku");

    when(settingsPane.fileChooser.showOpenDialog(any(Component.class))).thenReturn(APPROVE_OPTION);
    when(settingsPane.fileChooser.getSelectedFile()).thenReturn(file);
    when(settingsPane.configurationLoader.load(file)).thenReturn(config);
    when(settingsPane.language.getSelectedItem()).thenReturn("ja.hankaku");

    settingsPane.importConfig();

    verify(settingsPane.fileChooser).showOpenDialog(settingsPane.root);
    verify(settingsPane.fileChooser).getSelectedFile();
    verify(settingsPane.configurationLoader).load(file);
    assertSame(settingsPane.getConfig(), config);
    verify(settingsPane).initTabs();
    verify(settingsPane.language).setSelectedItem("ja.hankaku");
  }

  @Test
  public void canImportConfigurationAddingNewLanguage() throws Exception {
    prepareImportExport();
    Configuration config = config("za");
    Configuration clone1 = config("za");
    Configuration clone2 = config("za");

    when(settingsPane.fileChooser.showOpenDialog(any(Component.class))).thenReturn(APPROVE_OPTION);
    when(settingsPane.configurationLoader.load(any(File.class))).thenReturn(config);
    when(config.clone()).thenReturn(clone1, clone2);

    settingsPane.importConfig();

    assertSame(clone1, settingsPane.provider.getInitialConfig("za"));
    assertSame(clone2, settingsPane.provider.getConfig("za"));
    verify(settingsPane.language).addItem("za");
    verify(settingsPane.language, times(2)).setSelectedItem("za");
  }

  @SuppressWarnings("unchecked")
  private void prepareImportExport() {
    settingsPane.fileChooser = mock(JFileChooser.class);
    settingsPane.configurationLoader = mock(ConfigurationLoader.class);
    settingsPane.configurationExporter = mock(ConfigurationExporter.class);
    settingsPane.language = mock(JComboBox.class);
    doNothing().when(settingsPane).initTabs();
    doNothing().when(settingsPane).apply();
  }

  @Test
  public void applyValidators() throws Exception {
    List<ValidatorConfiguration> allValidators = asList(new ValidatorConfiguration("1"), new ValidatorConfiguration("2"));
    settingsPane.setConfig(configWithValidators(allValidators));

    List<ValidatorConfiguration> activeValidators = new ArrayList<>(allValidators.subList(0, 1));
    doReturn(activeValidators).when(settingsPane).getActiveValidators();

    settingsPane.applyValidatorsChanges();

    assertEquals(activeValidators, settingsPane.getConfig().getValidatorConfigs());
  }

  @Test
  public void applySymbols() throws Exception {
    Symbol symbol = new Symbol(BACKSLASH, '\\');
    doReturn(singletonList(symbol)).when(settingsPane).getSymbols();
    doReturn(mock(Configuration.class, RETURNS_DEEP_STUBS)).when(settingsPane).getConfig();

    settingsPane.applySymbolsChanges();

    verify(settingsPane.getConfig().getSymbolTable()).overrideSymbol(symbol);
  }

  @Test
  public void applyClonesLocalConfigs() throws Exception {
    doNothing().when(settingsPane).applyLocalChanges();
    settingsPane.apply();
    verify(settingsPane).applyLocalChanges();
    verify(settingsPane).cloneConfigs();
  }

  @Test
  public void resetToDefaults() throws Exception {
    settingsPane.initButtons();
    doNothing().when(settingsPane).initTabs();
    settingsPane.provider = mock(RedPenProvider.class);

    settingsPane.resetButton.doClick();

    verify(settingsPane.provider).reset();
    verify(settingsPane).cloneConfigs();
    verify(settingsPane).initTabs();
  }

  @Test
  public void resetChanges() throws Exception {
    doNothing().when(settingsPane).cloneConfigs();
    doNothing().when(settingsPane).initTabs();

    settingsPane.resetChanges();

    verify(settingsPane).cloneConfigs();
    verify(settingsPane).initTabs();
  }

  private ValidatorConfiguration validatorConfig(String name, Map<String, String> attributes) {
    ValidatorConfiguration config = new ValidatorConfiguration(name);
    attributes.entrySet().stream().forEach(entry -> config.addAttribute(entry.getKey(), entry.getValue()));
    return config;
  }
}

File: test/cc/redpen/intellij/SettingsPaneTest.kt
package cc.redpen.intellij

import cc.redpen.config.Configuration
import cc.redpen.config.Symbol
import cc.redpen.config.SymbolType.*
import cc.redpen.config.ValidatorConfiguration
import com.nhaarman.mockito_kotlin.*
import org.junit.Assert.*
import org.junit.Test
import org.mockito.Mockito.RETURNS_DEEP_STUBS
import java.io.File
import java.util.*
import java.util.Arrays.asList
import java.util.Collections.emptyMap
import javax.swing.CellEditor
import javax.swing.DefaultCellEditor
import javax.swing.JFileChooser.APPROVE_OPTION
import javax.swing.JFileChooser.CANCEL_OPTION
import javax.swing.JTextField
import javax.swing.event.ChangeEvent
import javax.swing.table.DefaultTableModel

class SettingsPaneTest : BaseTest() {
    var settingsPane: SettingsPane

    init {
        provider = RedPenProvider(project, LinkedHashMap(mapOf("en" to cloneableConfig("en"), "ja" to cloneableConfig("ja"))))
        settingsPane = spy(SettingsPane(provider))
        settingsPane.validators = mock(RETURNS_DEEP_STUBS)
        settingsPane.symbols = mock(RETURNS_DEEP_STUBS)
    }

    @Test
    fun allConfigsAreClonedOnCreation() {
        assertSame(settingsPane.provider.initialConfigs["en"]!!.clone(), settingsPane.configs["en"])
        assertSame(settingsPane.provider.initialConfigs["ja"]!!.clone(), settingsPane.configs["ja"])
    }

    @Test
    fun languagesAndVariantsArePrepopulated() {
        provider.activeConfig = provider.configs["ja"]!!

        settingsPane.initLanguages()

        assertEquals(2, settingsPane.language.itemCount.toLong())
        assertEquals("en", settingsPane.language.getItemAt(0))
        assertEquals("ja", settingsPane.language.getItemAt(1))
        assertEquals("ja", settingsPane.language.selectedItem)
    }

    @Test
    fun getPaneInitsEverything() {
        doNothing().whenever(settingsPane).initLanguages()
        doNothing().whenever(settingsPane).initValidators()
        doNothing().whenever(settingsPane).initSymbols()
        doNothing().whenever(settingsPane).initButtons()

        settingsPane.pane

        verify(settingsPane).initLanguages()
        verify(settingsPane).initValidators()
        verify(settingsPane).initSymbols()
        verify(settingsPane).initButtons()
    }

    @Test
    fun changingOfLanguageAppliesOldChangesAndInitsNewValidatorsAndSymbols() {
        doNothing().whenever(settingsPane).initTabs()
        doNothing().whenever(settingsPane).applyChanges()

        settingsPane.pane
        assertSame(provider.activeConfig.clone(), settingsPane.config)
        verify(settingsPane).initTabs()

        settingsPane.language.firePopupMenuWillBecomeVisible()
        verify(settingsPane).applyChanges()

        settingsPane.language.selectedItem = "ja"
        assertSame(provider.configs["ja"]!!.clone(), settingsPane.config)
        verify(settingsPane, times(2)).initTabs()
    }

    @Test
    fun validatorsAreListedInSettings() {
        val allValidators = asList(
                validatorConfig("ModifiedAttributes", mapOf("attr1" to "val1", "attr2" to "val2")),
                validatorConfig("InitialAttributes", mapOf("attr1" to "val1", "attr2" to "val2", "space" to " ")),
                validatorConfig("NoAttributes", emptyMap()))

        whenever(provider.initialConfigs["en"]!!.validatorConfigs).thenReturn(allValidators)
        doReturn(configWithValidators(listOf(
                validatorConfig("ModifiedAttributes", mapOf("foo" to "bar")),
                validatorConfig("NewValidator", mapOf("key" to "value"))))).whenever(settingsPane).config

        val model = mock<DefaultTableModel>()
        whenever(settingsPane.validators.model).thenReturn(model)

        settingsPane.initValidators()

        verify(model).addRow(arrayOf(true, "ModifiedAttributes", "foo=bar"))
        verify(model).addRow(arrayOf(false, "InitialAttributes", "attr2=val2; attr1=val1; space= "))
        verify(model).addRow(arrayOf(false, "NoAttributes", ""))
        verify(model).addRow(arrayOf(true, "NewValidator", "key=value"))
    }

    @Test
    fun getEditedValidators_returnsOnlySelectedValidators() {
        settingsPane.config = cloneableConfig("en")

        settingsPane.initLanguages()
        whenever(provider.initialConfigs["en"]!!.validatorConfigs).thenReturn(asList(
                ValidatorConfiguration("first"),
                ValidatorConfiguration("second one")))

        whenever(settingsPane.validators.model.rowCount).thenReturn(2)
        whenever(settingsPane.validators.model.getValueAt(0, 0)).thenReturn(false)
        whenever(settingsPane.validators.model.getValueAt(0, 1)).thenReturn("first")
        whenever(settingsPane.validators.model.getValueAt(1, 0)).thenReturn(true)
        whenever(settingsPane.validators.model.getValueAt(1, 1)).thenReturn("second one")
        whenever(settingsPane.validators.model.getValueAt(1, 2)).thenReturn("")

        val activeValidators = settingsPane.getEditedValidators()
        assertEquals(1, activeValidators.size.toLong())
        assertEquals("second one", activeValidators[0].configurationName)
    }

    @Test
    fun getEditedValidators_modifiesAttributes() {
        settingsPane.config = cloneableConfig("en")
        settingsPane.initLanguages()
        whenever(provider.initialConfigs["en"]!!.validatorConfigs).thenReturn(
                listOf(validatorConfig("Hello", mapOf("width" to "100", "height" to "300", "depth" to "1"))))

        whenever(settingsPane.validators.model.rowCount).thenReturn(1)
        whenever(settingsPane.validators.model.getValueAt(0, 0)).thenReturn(true)
        whenever(settingsPane.validators.model.getValueAt(0, 1)).thenReturn("Hello")
        whenever(settingsPane.validators.model.getValueAt(0, 2)).thenReturn(" width=200;   height=300; space= ")

        val activeValidators = settingsPane.getEditedValidators()
        assertEquals(1, activeValidators.size.toLong())
        assertEquals(mapOf("width" to "200", "height" to "300", "space" to " "), activeValidators[0].properties)
        assertNotSame(provider.initialConfigs["en"]!!.validatorConfigs[0], activeValidators[0])
    }

    @Test
    fun getEditedValidators_doesNotApplyActiveCellEditorChanges() {
        settingsPane.config = cloneableConfig("en")
        whenever(settingsPane.validators.isEditing).thenReturn(true)
        settingsPane.getEditedValidators()
        verify(settingsPane.validators.cellEditor, never()).stopCellEditing()
    }

    @Test
    fun getEditedValidators_usesCurrentlyOpenCellEditor() {
        settingsPane.config = cloneableConfig("en")
        settingsPane.initLanguages()
        whenever(provider.initialConfigs["en"]!!.validatorConfigs).thenReturn(
                listOf(validatorConfig("Hello", emptyMap())))

        whenever(settingsPane.validators.model.rowCount).thenReturn(1)
        whenever(settingsPane.validators.model.getValueAt(0, 0)).thenReturn(true)
        whenever(settingsPane.validators.model.getValueAt(0, 1)).thenReturn("Hello")

        val cellEditor = mock<DefaultCellEditor>()
        whenever(settingsPane.validators.cellEditor).thenReturn(cellEditor)
        val cellEditorField = mock<JTextField>()
        whenever(cellEditor.component).thenReturn(cellEditorField)
        whenever(cellEditorField.text).thenReturn("foo=bar")
        whenever(settingsPane.validators.editingRow).thenReturn(0)

        val activeValidators = settingsPane.getEditedValidators()

        assertEquals(mapOf("foo" to "bar"), activeValidators[0].properties)
    }

    @Test
    fun getEditedSymbols_doesNotApplyActiveCellEditorChanges() {
        whenever(settingsPane.symbols.isEditing).thenReturn(true)
        settingsPane.getEditedSymbols()
        verify(settingsPane.symbols.cellEditor, never()).stopCellEditing()
    }

    @Test
    fun symbolsAreListedInSettings() {
        settingsPane.config = configWithSymbols(asList(Symbol(AMPERSAND, '&', "$%", true, false), Symbol(ASTERISK, '*', "", false, true)))

        val model = mock<DefaultTableModel>()
        whenever(settingsPane.symbols.model).thenReturn(model)

        settingsPane.initSymbols()

        verify(model).addRow(arrayOf(AMPERSAND.toString(), '&', "$%", true, false))
        verify(model).addRow(arrayOf(ASTERISK.toString(), '*', "", false, true))
    }

    @Test
    fun getEditedSymbols() {
        val model = settingsPane.symbols.model
        whenever(model.rowCount).thenReturn(2)

        whenever(model.getValueAt(0, 0)).thenReturn("AMPERSAND")
        whenever(model.getValueAt(0, 1)).thenReturn('&')
        whenever(model.getValueAt(0, 2)).thenReturn("$%")
        whenever(model.getValueAt(0, 3)).thenReturn(true)
        whenever(model.getValueAt(0, 4)).thenReturn(false)

        whenever(model.getValueAt(1, 0)).thenReturn("ASTERISK")
        whenever(model.getValueAt(1, 1)).thenReturn("*")
        whenever(model.getValueAt(1, 2)).thenReturn("")
        whenever(model.getValueAt(1, 3)).thenReturn(false)
        whenever(model.getValueAt(1, 4)).thenReturn(true)

        val symbols = settingsPane.getEditedSymbols()
        assertEquals(asList(Symbol(AMPERSAND, '&', "$%", true, false), Symbol(ASTERISK, '*', "", false, true)), symbols)
    }


    @Test
    fun getEditedSymbols_usesCurrentlyOpenCellEditor() {
        val model = settingsPane.symbols.model
        whenever(model.rowCount).thenReturn(1)

        whenever(model.getValueAt(0, 0)).thenReturn("AMPERSAND")
        whenever(model.getValueAt(0, 1)).thenReturn('&')
        whenever(model.getValueAt(0, 2)).thenReturn("%*")
        whenever(model.getValueAt(0, 3)).thenReturn(true)
        whenever(model.getValueAt(0, 4)).thenReturn(false)

        val cellEditor = mock<DefaultCellEditor>()
        whenever(settingsPane.symbols.cellEditor).thenReturn(cellEditor)
        val cellEditorField = mock<JTextField>()
        whenever(cellEditor.component).thenReturn(cellEditorField)
        whenever(cellEditorField.text).thenReturn("$")
        whenever(settingsPane.symbols.editingRow).thenReturn(0)
        whenever(settingsPane.symbols.editingColumn).thenReturn(1)

        assertEquals('$', settingsPane.getEditedSymbols()[0].value)

        whenever(cellEditorField.text).thenReturn("abc")
        whenever(settingsPane.symbols.editingColumn).thenReturn(2)

        assertArrayEquals(charArrayOf('a', 'b', 'c'), settingsPane.getEditedSymbols()[0].invalidChars)
    }

    @Test
    fun fileChooserUsesXmlFileFilter() {
        assertEquals("RedPen Configuration", settingsPane.fileChooser.fileFilter.description)
        val file = mock<File>()

        whenever(file.name).thenReturn("blah.xml")
        assertTrue(settingsPane.fileChooser.fileFilter.accept(file))

        whenever(file.name).thenReturn("blah.txt")
        assertFalse(settingsPane.fileChooser.fileFilter.accept(file))
    }

    @Test
    fun canCancelExportingConfiguration() {
        prepareImportExport()
        settingsPane.initButtons()
        whenever(settingsPane.fileChooser.showSaveDialog(any())).thenReturn(CANCEL_OPTION)

        settingsPane.exportButton.doClick()

        verify(settingsPane, never()).save()
        verify(settingsPane.fileChooser).showSaveDialog(settingsPane.root)
        verifyNoMoreInteractions(settingsPane.fileChooser)
    }

    @Test
    fun canExportConfiguration() {
        prepareImportExport()
        val file = File.createTempFile("redpen-conf", ".xml")
        file.deleteOnExit()

        doReturn(config("en")).whenever(settingsPane).config
        whenever(settingsPane.fileChooser.showSaveDialog(any())).thenReturn(APPROVE_OPTION)
        whenever(settingsPane.fileChooser.selectedFile).thenReturn(file)

        settingsPane.exportConfig()

        verify(settingsPane).save()
        verify(settingsPane.fileChooser).showSaveDialog(settingsPane.root)
        verify(settingsPane.fileChooser).selectedFile
        verify(settingsPane.configurationExporter).export(eq(settingsPane.config), any())
    }

    @Test
    fun canCancelImportingConfiguration() {
        prepareImportExport()
        settingsPane.initButtons()
        whenever(settingsPane.fileChooser.showOpenDialog(any())).thenReturn(CANCEL_OPTION)

        settingsPane.importButton.doClick()

        verify(settingsPane.fileChooser).showOpenDialog(settingsPane.root)
        verifyNoMoreInteractions(settingsPane.fileChooser)
    }

    @Test
    fun canImportConfiguration() {
        prepareImportExport()
        val file = mock<File>()
        val config = config("ja.hankaku")

        whenever(settingsPane.fileChooser.showOpenDialog(any())).thenReturn(APPROVE_OPTION)
        whenever(settingsPane.fileChooser.selectedFile).thenReturn(file)
        whenever(settingsPane.configurationLoader.load(file)).thenReturn(config)
        whenever(settingsPane.language.selectedItem).thenReturn("ja.hankaku")

        settingsPane.importConfig()

        verify(settingsPane.fileChooser).showOpenDialog(settingsPane.root)
        verify(settingsPane.fileChooser).selectedFile
        verify(settingsPane.configurationLoader).load(file)
        assertSame(settingsPane.config, config)
        verify(settingsPane).initTabs()
        verify(settingsPane.language).selectedItem = "ja.hankaku"
    }

    @Test
    fun canImportConfigurationAddingNewLanguage() {
        prepareImportExport()
        val config = config("za")
        val clone1 = config("za")
        val clone2 = config("za")

        whenever(settingsPane.fileChooser.showOpenDialog(any())).thenReturn(APPROVE_OPTION)
        whenever(settingsPane.configurationLoader.load(any<File>())).thenReturn(config)
        whenever(config.clone()).thenReturn(clone1, clone2)

        settingsPane.importConfig()

        assertSame(clone1, settingsPane.provider.initialConfigs["za"])
        assertSame(clone2, settingsPane.provider.configs["za"])
        verify(settingsPane.language).addItem("za")
        verify(settingsPane.language, times(2)).selectedItem = "za"
    }

    private fun prepareImportExport() {
        settingsPane.fileChooser = mock()
        settingsPane.configurationLoader = mock()
        settingsPane.configurationExporter = mock()
        settingsPane.language = mock()
        doNothing().whenever(settingsPane).initTabs()
        doNothing().whenever(settingsPane).save()
    }

    @Test
    fun applyValidatorChanges() {
        val allValidators = asList(ValidatorConfiguration("1"), ValidatorConfiguration("2"))
        settingsPane.config = configWithValidators(allValidators)

        val activeValidators = ArrayList(allValidators.subList(0, 1))
        doReturn(activeValidators).whenever(settingsPane).getEditedValidators()

        settingsPane.applyValidatorsChanges()

        assertEquals(activeValidators, settingsPane.config.validatorConfigs)
    }

    @Test
    fun applyValidatorChanges_addsNewValidatorsIfNeeded() {
        whenever(provider.initialConfigs["en"]!!.validatorConfigs).thenReturn(asList(
                ValidatorConfiguration("1"),
                ValidatorConfiguration("2")))
        val validators = asList(
                ValidatorConfiguration("2"),
                ValidatorConfiguration("active new"),
                ValidatorConfiguration("inactive new"))

        doReturn(configWithValidators(validators)).whenever(settingsPane).config

        whenever(settingsPane.validators.model.rowCount).thenReturn(2)
        whenever(settingsPane.validators.model.getValueAt(0, 0)).thenReturn(true)
        whenever(settingsPane.validators.model.getValueAt(0, 1)).thenReturn("2")
        whenever(settingsPane.validators.model.getValueAt(0, 2)).thenReturn("")

        whenever(settingsPane.validators.model.getValueAt(1, 0)).thenReturn(true)
        whenever(settingsPane.validators.model.getValueAt(1, 1)).thenReturn("active new")
        whenever(settingsPane.validators.model.getValueAt(1, 2)).thenReturn("")

        whenever(settingsPane.validators.model.getValueAt(2, 0)).thenReturn(false)
        whenever(settingsPane.validators.model.getValueAt(2, 1)).thenReturn("inactive new")
        whenever(settingsPane.validators.model.getValueAt(2, 2)).thenReturn("")

        settingsPane.applyValidatorsChanges()

        assertEquals(validators.subList(0, 2), settingsPane.config.validatorConfigs)
    }

    @Test
    fun applySymbolsChanges() {
        val symbol = Symbol(BACKSLASH, '\\')
        doReturn(listOf(symbol)).whenever(settingsPane).getEditedSymbols()
        doReturn(mock<Configuration>(RETURNS_DEEP_STUBS)).whenever(settingsPane).config

        settingsPane.applySymbolsChanges()

        verify(settingsPane.config.symbolTable).overrideSymbol(symbol)
    }

    @Test
    fun saveClonesLocalConfigs() {
        doNothing().whenever(settingsPane).applyChanges()
        doNothing().whenever(settingsPane).cloneConfigs()
        settingsPane.save()
        verify(settingsPane).applyChanges()
        verify(settingsPane).cloneConfigs()
    }

    @Test
    fun resetToDefaults() {
        settingsPane.initButtons()
        doNothing().whenever(settingsPane).initTabs()

        settingsPane.resetButton.doClick()

        assertEquals(settingsPane.provider.initialConfigs["en"]!!.clone(), settingsPane.configs["en"])
        assertEquals(settingsPane.provider.initialConfigs["ja"]!!.clone(), settingsPane.configs["ja"])
        verify(settingsPane).initTabs()
    }

    @Test
    fun resetChanges() {
        doNothing().whenever(settingsPane).cloneConfigs()
        doNothing().whenever(settingsPane).initTabs()

        settingsPane.resetChanges()

        verify(settingsPane).cloneConfigs()
        verify(settingsPane).initTabs()
    }

    @Test
    fun applyChanges() {
        doNothing().whenever(settingsPane).applySymbolsChanges()
        doNothing().whenever(settingsPane).applyValidatorsChanges()

        settingsPane.applyChanges()

        verify(settingsPane).applySymbolsChanges()
        verify(settingsPane).applyValidatorsChanges()
    }

    @Test
    fun isCorrectPropertiesFormat() {
        assertTrue(settingsPane.isCorrectValidatorPropertiesFormat(""))
        assertTrue(settingsPane.isCorrectValidatorPropertiesFormat("foo="))
        assertTrue(settingsPane.isCorrectValidatorPropertiesFormat("foo=bar"))
        assertTrue(settingsPane.isCorrectValidatorPropertiesFormat("foo=bar;"))
        assertTrue(settingsPane.isCorrectValidatorPropertiesFormat("foo=bar;foo2="))
        assertTrue(settingsPane.isCorrectValidatorPropertiesFormat("foo=bar;foo2=bar2"))
        assertTrue(settingsPane.isCorrectValidatorPropertiesFormat("foo=bar;foo2=bar2;"))
        assertTrue(settingsPane.isCorrectValidatorPropertiesFormat("foo=bar;foo2=bar2;foo3=bar3"))
        assertFalse(settingsPane.isCorrectValidatorPropertiesFormat("foo"))
        assertFalse(settingsPane.isCorrectValidatorPropertiesFormat("=bar"))
        assertFalse(settingsPane.isCorrectValidatorPropertiesFormat("=bar;"))
        assertFalse(settingsPane.isCorrectValidatorPropertiesFormat("foo=bar;="))
        assertFalse(settingsPane.isCorrectValidatorPropertiesFormat("foo=bar;foo2"))
    }

    @Test
    fun reportInvalidValidatorAttributesFormatWhenEditorIsStopped() {
        doNothing().whenever(settingsPane).showValidatorPropertyError(any())
        val event = mock<ChangeEvent>()
        val source = mock<CellEditor>()
        whenever(event.source).thenReturn(source)

        whenever(source.cellEditorValue).thenReturn("width")
        settingsPane.showValidatorPropertyErrorIfNeeded(event)
        verify(settingsPane).showValidatorPropertyError("width")

        whenever(source.cellEditorValue).thenReturn("=")
        settingsPane.showValidatorPropertyErrorIfNeeded(event)
        verify(settingsPane).showValidatorPropertyError("=")

        whenever(source.cellEditorValue).thenReturn("width=120")
        settingsPane.showValidatorPropertyErrorIfNeeded(event)
        verifyNoMoreInteractions(settingsPane)
    }

    private fun validatorConfig(name: String, attributes: Map<String, String>): ValidatorConfiguration {
        val config = ValidatorConfiguration(name)
        attributes.forEach{ e -> config.addProperty(e.key, e.value) }
        return config
    }
}